middle_end/flambda2/types/traversals.ml

```diff
@@ -0,0 +1,1335 @@
+
+  and rewrite_head_of_kind_naked_vec512 head : _ Or_unknown.t =
+    Or_unknown.Known head
+
+  and rewrite_head_of_kind_rec_info head : _ Or_unknown.t =
+    Or_unknown.Known head
+
+  and rewrite_head_of_kind_region () : _ Or_unknown.t = Or_unknown.Known ()
+
+  and rewrite_row_like_for_blocks env acc metadata
+      ({ known_tags; other_tags; alloc_mode } : TG.row_like_for_blocks) =
+    let known_tags, acc =
+      Tag.Map.fold
+        (fun tag case (known_tags, acc) ->
+          let case, acc =
+            match (case : _ TG.row_like_case Or_unknown.t) with
+            | Unknown -> case, acc
+            | Known { maps_to; env_extension = _; index } ->
+              let maps_to, acc =
+                rewrite_int_indexed_product
+                  ~slot:(fun t index env ty -> X.block_slot ~tag t index env ty)
+                  env acc metadata maps_to
+              in
+              (* Drop env extension because it is not clear what to do. *)
+              ( Or_unknown.Known
+                  (TG.Row_like_case.create ~maps_to
+                     ~env_extension:TG.Env_extension.empty ~index),
+                acc )
+          in
+          Tag.Map.add tag case known_tags, acc)
+        known_tags (Tag.Map.empty, acc)
+    in
+    let other_tags, acc =
+      match (other_tags : _ TG.row_like_case Or_bottom.t) with
+      | Bottom -> Or_bottom.Bottom, acc
+      | Ok { maps_to; env_extension = _; index } ->
+        let maps_to, acc =
+          rewrite_int_indexed_product
+            ~slot:(fun t index env ty -> X.block_slot t index env ty)
+            env acc metadata maps_to
+        in
+        ( Or_bottom.Ok
+            (TG.Row_like_case.create ~maps_to
+               ~env_extension:TG.Env_extension.empty ~index),
+          acc )
+    in
+    TG.Row_like_for_blocks.create_raw ~known_tags ~other_tags ~alloc_mode, acc
+
+  and rewrite_row_like_for_closures env acc metadata
+      ({ known_closures; other_closures } : TG.row_like_for_closures) =
+    let known_closures, acc =
+      Function_slot.Map.fold
+        (fun function_slot
+             ({ maps_to; env_extension = _; index } : _ TG.row_like_case)
+             (known_closures, acc) ->
+          let set_of_closures_metadata =
+            X.set_of_closures metadata function_slot env maps_to
+          in
+          let maps_to, acc =
+            rewrite_closures_entry ~this_function_slot:function_slot env acc
+              set_of_closures_metadata maps_to
+          in
+          let row_like_case =
+            TG.Row_like_case.create ~maps_to
+              ~env_extension:TG.Env_extension.empty ~index
+          in
+          Function_slot.Map.add function_slot row_like_case known_closures, acc)
+        known_closures
+        (Function_slot.Map.empty, acc)
+    in
+    let other_closures, acc =
+      match other_closures with
+      | Bottom -> Or_bottom.Bottom, acc
+      | Ok { maps_to = _; env_extension = _; index = _ } -> assert false
```

Comment:
I think this deserves a comment (i.e. we should remove the `other_closures` field completely, as it is always Bottom, but not in this PR).